reframe/utility/typecheck.py

```diff
@@ -322,19 +323,15 @@ def __rfm_cast_str__(cls, s):
         mappping_type = cls._type
         key_type = cls._key_type
         value_type = cls._value_type
-        seq = []
-        for key_datum in s.split(','):
-            try:
-                k, v = key_datum.split(':')
-            except ValueError:
-                # Re-raise as TypeError
-                raise TypeError(
-                    f'cannot convert string {s!r} to {cls.__name__!r}'
-                ) from None
-
-            seq.append((key_type(k), value_type(v)))
-
-        return mappping_type(seq)
+        # Convert string to JSON format: wrap it with braces and add double
+        # quotes where necessary.
+        s = f'{{{s}}}'
+        s = re.sub('([:,{] *)([^":,{}]+)', r'\1"\2"', s)
+        # Parse the JSON string.
+        d = json.loads(s)
+        # Enforce the desired output types.
+        d = {key_type(key): value_type(value) for key, value in d.items()}
```

Comment:
I added this to force conversion to the desired types and make existing tests pass, but this now breaks parsing of `extra_resources`, I get again `type error: failed to set field 'extra_resources': could not convert to any of the supported types: ['Dict[str,Dict[str,object]]']` :smile:

Reply:
Uhm, something like
```suggestion
        if not issubclass(value_type, object):
            d = {key_type(key): value_type(value) for key, value in d.items()}
```
should do the trick.  Does it make sense?

Reply:
I don't think we should do any manipulation of the syntax here. What if we first tried to convert from json and if that failed, fall back to the current conversion? So, for a normal dict, those two syntaxes would be equivalent:

1. `-S env_vars=OMP_NUM_THREADS:4`
2. `-S env_vars='{"OMP_NUM_THREADS": "4"}'`

For the nested dicts one should write always valid json: `-S extra_resources='{"gpu":{"num_gpus_per_node":8}}'`. 

We should do the same for the other aggregate types, too.

Reply:
I was just trying to support current brace- and quotes-less syntax, but if you prefer valid JSON syntax and fallback on current syntax, that'd be good for me.

Reply:
Ok, I changed this PR to first try to parse with json and then fall back to the old parsing strategy unchanged, if `json.loads` fails.  I also added a test.